
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../data/">
      
      
        <link rel="next" href="../cli/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.30">
    
    
      
        <title>Package Overview - TRIBAL</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.3cba04c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="brown" data-md-color-accent="deep-purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#package-overview" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="TRIBAL" class="md-header__button md-logo" aria-label="TRIBAL" data-md-component="logo">
      
  <img src="../../img/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TRIBAL
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Package Overview
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/elkebir-group/TRIBAL" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  About

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../install/" class="md-tabs__link">
          
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../api/preprocess/" class="md-tabs__link">
          
  
  API

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="TRIBAL" class="md-nav__button md-logo" aria-label="TRIBAL" data-md-component="logo">
      
  <img src="../../img/logo.png" alt="logo">

    </a>
    TRIBAL
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/elkebir-group/TRIBAL" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    About
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Usage
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download & Installation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Package Overview
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Package Overview
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example-data" class="md-nav__link">
    <span class="md-ellipsis">
      Example data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#using-the-package" class="md-nav__link">
    <span class="md-ellipsis">
      Using the package
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Using the package">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocessing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#running-tribal" class="md-nav__link">
    <span class="md-ellipsis">
      Running TRIBAL
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#exploring-and-visualizing-the-inferred-b-cell-lineage-trees" class="md-nav__link">
    <span class="md-ellipsis">
      Exploring and visualizing the inferred B cell lineage trees
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#functionality-of-a-lineagetreelist" class="md-nav__link">
    <span class="md-ellipsis">
      Functionality of a LineageTreeList
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Command Line Interface
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Preprocesssing
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/base_tree/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    BaseTree
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/clonotype/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clonotype
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/tribal/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tribal
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/lineagetree/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LineageTree
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/lineagetreelist/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LineageTreeList
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="package-overview">Package overview<a class="headerlink" href="#package-overview" title="Permanent link">&para;</a></h1>
<p>The <code>tribal</code> package can be imported as a package into a python script or jupyter notebookd or it can be used as a <a href="../cli/">command line tool</a>. </p>
<p>The following functions and classes are accessible via the <code>tribal</code> package:  </p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Type</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../../api/preprocess/">preprocess</a></td>
<td>preprocess the input data to the correct format for tribal by finding a multiple sequence aligment and parsimony forest for each clonotype</td>
<td>function</td>
</tr>
<tr>
<td><a href="../../api/base_tree/">BaseTree</a></td>
<td>a class to model the lineage tree topology</td>
<td>class</td>
</tr>
<tr>
<td><a href="../../api/clonotype/">Clonotype</a></td>
<td>a dataclass to structure the input data for <code>tribal</code></td>
<td>class</td>
</tr>
<tr>
<td><a href="../../api/tribal/">Tribal</a></td>
<td>the main class to run the <code>tribal</code> algorithm and fit the input data</td>
<td>class</td>
</tr>
<tr>
<td><a href="../../api/lineagetree/">LineageTree</a></td>
<td>a class to model an inferred B cell lineage tree</td>
<td>class</td>
</tr>
<tr>
<td><a href="../../api/lineagetree/">LineageTreeList</a></td>
<td>an extensions of class list to contain a list of B cell lineage trees</td>
<td>class</td>
</tr>
</tbody>
</table>
<p>The API provides additional details on each of these items. </p>
<h2 id="example-data">Example data<a class="headerlink" href="#example-data" title="Permanent link">&para;</a></h2>
<p>In addition to the above functions and class, the following example data can be imported to help users better understand the data formatting and package use. </p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
<th>Type</th>
</tr>
</thead>
<tbody>
<tr>
<td>df</td>
<td>input sequencing data</td>
<td><code>pandas.DataFrame</code></td>
</tr>
<tr>
<td>roots</td>
<td>input germline roots for sequencing data</td>
<td><code>pandas.DataFrame</code></td>
</tr>
<tr>
<td>probabilities</td>
<td>example isotype transition probability matrix</td>
<td><code>numpy.ndarray</code></td>
</tr>
<tr>
<td>clonotypes</td>
<td>dictionary of <a href="../../api/clonotype/">Clonotype</a> objects</td>
<td><code>dict</code></td>
</tr>
<tr>
<td>lineage_tree</td>
<td>an example inferred <a href="../../api/lineagetree/">B cell lineage tree</a></td>
<td><a href="../../api/lineagetree/">LineageTree</a></td>
</tr>
<tr>
<td>lineage_tree_list</td>
<td>an example inferred <a href="../../api/lineagetreelist/">B cell lineage tree list</a></td>
<td><a href="../../api/lineagetree/">LineageTreeList</a></td>
</tr>
</tbody>
</table>
<p>See <a href="../data/">Data</a> for more details on the input format for the data. </p>
<p>Load and view the example input data:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">df</span><span class="p">,</span> <span class="n">roots</span>

<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">roots</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</code></pre></div>
<p>Load and view the example output data from <code>preprocess</code>:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">clonotypes</span>
<span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">clonotype</span> <span class="ow">in</span> <span class="n">clonptypes</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">clonotype</span><span class="p">)</span>
</code></pre></div>
<p>Load and view the example output data from the <code>tribal</code> algorithm:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">probabilities</span><span class="p">,</span> <span class="n">lineage_tree</span><span class="p">,</span> <span class="n">lineage_tree_list</span>
<span class="nb">print</span><span class="p">(</span><span class="n">probabilities</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree_list</span><span class="p">)</span>
</code></pre></div>
<h2 id="using-the-package">Using the package<a class="headerlink" href="#using-the-package" title="Permanent link">&para;</a></h2>
<p>Here is a brief walkthrough of how to utilize the functionality of the <code>tribal</code> package.
First, load the package:</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span> <span class="nn">tribal</span>
</code></pre></div>
<p>or, alternatively load specific functions, classes or example data.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">preprocess</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">roots</span>
</code></pre></div>
<h3 id="preprocessing">Preprocessing<a class="headerlink" href="#preprocessing" title="Permanent link">&para;</a></h3>
<p>The <a href="../../api/preprocess/">preprocess</a> function will:<br />
    -  filter out clonotypes that are below the minimum number of cells .
    -  filter out cells which have v alleles that differ from the majority of the clonotype<br />
    -  perform a multiple sequence alignment (MSA) for each valid clonotype using <a href="https://mafft.cbrc.jp/alignment/software/">mafft</a><br />
    -  infer a parsimony forest for each clonotype given the MSA using <a href="https://phylipweb.github.io/phylip/">dnapars</a>  </p>
<p>See <a href="../../api/preprocess/">preprocess</a> for more details. </p>
<div class="highlight"><pre><span></span><code>    <span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">preprocess</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="n">roots</span>
    <span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>
    <span class="n">clonotypes</span><span class="p">,</span> <span class="n">df_filt</span> <span class="o">=</span> <span class="n">preprocess</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">roots</span><span class="p">,</span><span class="n">isotypes</span><span class="o">=</span><span class="n">isotypes</span><span class="p">,</span> 
                                    <span class="n">min_size</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">use_light_chain</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
                                    <span class="n">cores</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
<p>The output dictionary <code>clonotypes</code> is the formatted input to <code>tribal</code>.  To view
the formatted example data without running the preprocessing step, run the following.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">clonotypes</span>
<span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">clonotype</span> <span class="ow">in</span> <span class="n">clonotypes</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">clonotype</span><span class="p">)</span>
</code></pre></div>
<h3 id="running-tribal">Running TRIBAL<a class="headerlink" href="#running-tribal" title="Permanent link">&para;</a></h3>
<p>Tribal takes the dictionary of <a href="../../api/clonotype/">Clonotype</a> objects as input and can be run in two modes. <br />
1. <code>refinement</code> (recommended) : the full algorithm where the CSR likelihood is optimized by solving the most parsimonious tree refinement problem.<br />
2. <code>score</code> : the input parsimony lineage trees are not refined and isotypes of the internal nodes are inferred using weighted parsimony via the Sankoff algorithm, with the isotype transition probabilities as weights.   </p>
<p><div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">Tribal</span><span class="p">,</span> <span class="n">clonotypes</span>

<span class="c1">#the clonotype data contains the following isotypes encoded from 0 to 7</span>
<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>
<span class="n">tr</span> <span class="o">=</span> <span class="n">Tribal</span><span class="p">(</span><span class="n">n_isotypes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">isotypes</span><span class="p">),</span> <span class="n">restarts</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">niter</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1">#run in refinement mode (recommended)</span>
<span class="n">shm_score</span><span class="p">,</span> <span class="n">csr_likelihood</span><span class="p">,</span> <span class="n">best_trees</span><span class="p">,</span> <span class="n">probabilities</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">clonotypes</span><span class="o">=</span><span class="n">clonotypes</span><span class="p">,</span> 
                                                                <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;refinement&quot;</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1">#run in score mode to infer isotypes using weighted parsimony (Sankoff algorithm) w/o tree refinement</span>
<span class="n">shm_score</span><span class="p">,</span> <span class="n">csr_likelihood</span><span class="p">,</span> <span class="n">best_trees</span><span class="p">,</span> <span class="n">probabilities</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">clonotypes</span><span class="o">=</span><span class="n">clonotypes</span><span class="p">,</span> 
                                                                <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;score&quot;</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div>
<code>shm_score</code> and <code>csr_likelihood</code> are floats representing the corresponding SHM or CSR objective values. </p>
<p><code>probabilities</code> is a numpy array of shape <code>(n_isotypes, n_isotypes)</code> containing the inferred isotype transition probabilites. </p>
<p><code>best_trees</code> is a dictionary with clonotype id as key and the value containing a <a href="../../api/lineagetreelist/">LineageTreeList</a> with all inferred optimal B cell lineage trees for a given clonotype. </p>
<p>Additionally, <code>Tribal</code> can be <code>fit</code> with a user-provided isotype transition probability matrix:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">Tribal</span><span class="p">,</span> <span class="n">clonotypes</span><span class="p">,</span> <span class="n">probabilities</span>

<span class="c1">#the clonotype data contains the following isotypes encoded from 0 to 7</span>
<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>
<span class="n">tr</span> <span class="o">=</span> <span class="n">Tribal</span><span class="p">(</span><span class="n">n_isotypes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">isotypes</span><span class="p">),</span> <span class="n">restarts</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">niter</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1">#specifying the transmat argument will skip the step of inferring isotype transition probabilites</span>
<span class="n">shm_score</span><span class="p">,</span> <span class="n">csr_likelihood</span><span class="p">,</span> <span class="n">best_trees</span><span class="p">,</span> <span class="n">probabilities</span> <span class="o">=</span> <span class="n">tr</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">clonotypes</span><span class="o">=</span><span class="n">clonotypes</span><span class="p">,</span>
                                                                <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;refinement&quot;</span><span class="p">,</span> <span class="n">transmat</span><span class="o">=</span><span class="n">probabilites</span><span class="p">,</span> 
                                                                <span class="n">cores</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div>
<h3 id="exploring-and-visualizing-the-inferred-b-cell-lineage-trees">Exploring and visualizing the inferred B cell lineage trees<a class="headerlink" href="#exploring-and-visualizing-the-inferred-b-cell-lineage-trees" title="Permanent link">&para;</a></h3>
<p><code>tribal fit</code> returns a list of all optimal B cell lineage trees for each clonotype. 
Specifically, in the above examples <code>best_trees</code> is a dictionary, with clonotype as key, of <a href="../../api/lineagetreelist/">LineageTreeLists</a>. </p>
<p>A B cell lineage tree for tribal is a rooted tree with nodes labeled by BCR sequences (concentated heavy and optional light chains) and by isotypes. The <a href="../../api/lineagetree/">LineageTree</a> class also holds the current  SHM parsimony score (<code>shm_obj</code>) and CSR likelihood (<code>csr_obj</code>). </p>
<p>A <code>LineageTree</code> can be visualized as a <code>png</code> or <code>pdf</code> via the <a href="../api/lineage_tree.md">draw</a> function.  Nodes are colored by the isotype via the default <code>color_encoding</code>. </p>
<div class="highlight"><pre><span></span><code> <span class="n">color_encoding</span> <span class="o">=</span>  <span class="p">{</span>
                <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="s2">&quot;#FFFFFF&quot;</span><span class="p">,</span>
                <span class="mi">0</span> <span class="p">:</span> <span class="s2">&quot;#808080&quot;</span><span class="p">,</span>
                <span class="mi">1</span> <span class="p">:</span> <span class="s2">&quot;#FFEDA0&quot;</span><span class="p">,</span>
                <span class="mi">2</span> <span class="p">:</span> <span class="s2">&quot;#FD8D3C&quot;</span><span class="p">,</span>
                <span class="mi">3</span> <span class="p">:</span> <span class="s2">&quot;#E31A1C&quot;</span><span class="p">,</span>
                <span class="mi">4</span> <span class="p">:</span> <span class="s2">&quot;#800026&quot;</span><span class="p">,</span>
                <span class="mi">5</span> <span class="p">:</span> <span class="s2">&quot;#6A51A3&quot;</span><span class="p">,</span>
                <span class="mi">6</span> <span class="p">:</span> <span class="s2">&quot;#74C476&quot;</span><span class="p">,</span>
                <span class="mi">7</span> <span class="p">:</span> <span class="s2">&quot;mediumseagreen&quot;</span><span class="p">,</span>
                <span class="mi">8</span> <span class="p">:</span> <span class="s2">&quot;darkgoldenrod&quot;</span><span class="p">,</span>
                <span class="mi">9</span> <span class="p">:</span> <span class="s2">&quot;thistle1&quot;</span>
        <span class="p">}</span>
</code></pre></div>
<p><code>show_legend=True</code> provides a legend on the visualization depicting the color encoding. if <code>isotype_encoding</code> is provided in the form of a list, i.e., <code>['IGHM', 'IGHG3', 'IGHG1', 'IGHA1','IGHG2','IGHG4','IGHE','IGHA2']</code> then the legend will be utilize the isotype labeling. Otherwise, the encoding is used. </p>
<p>The <code>show_labels</code> argument can be used to toggle on and off the labeling of the sequences. </p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="p">)</span>

<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>
<span class="c1">#output visualization as a png with sequence labels included.</span>
<span class="n">lineage_tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">fname</span><span class="o">=</span><span class="s2">&quot;example_tree.png&quot;</span><span class="p">,</span> 
                <span class="n">isotype_encoding</span><span class="o">=</span><span class="n">isotypes</span><span class="p">,</span>
                <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">show_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">color_encoding</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>

<span class="c1">#output visualization as a pdf with sequence labels excluded.</span>
<span class="n">lineage_tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">fname</span><span class="o">=</span><span class="s2">&quot;example_tree.pdf&quot;</span><span class="p">,</span> 
                <span class="n">isotype_encoding</span><span class="o">=</span><span class="n">isotypes</span><span class="p">,</span>
                <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">show_labels</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">color_encoding</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>
</code></pre></div>
<p>The output file can also be saved as dot file instead of a png for pdf. Use the <code>dot</code> argument to indicate that file should be written as a dot file.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span>


<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>
<span class="c1">#output visualization as a png with sequence labels included.</span>
<span class="n">lineage_tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">fname</span><span class="o">=</span><span class="s2">&quot;example_tree.dot&quot;</span><span class="p">,</span> 
                <span class="n">isotype_encoding</span><span class="o">=</span><span class="n">isotypes</span><span class="p">,</span>
                <span class="n">show_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">show_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">color_encoding</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">dot</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="p">)</span>
</code></pre></div>
<p>Use the <a href="../../api/lineagetree/">write</a> function the lineage tree data to files including:<br />
1. The sequences as a FASTA file<br />
2. The isotypes as a csv file<br />
3. The tree as a png<br />
4. The edge list of the lineage tree</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span>
<span class="n">lineage_tree</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;lineage_tree_files&quot;</span><span class="p">)</span>
</code></pre></div>
<p>You can also pass the corresponding <code>Clonotype</code> object to utilize the stored isotype encoding for the clonotype.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span><span class="p">,</span> <span class="n">clonotypes</span>
<span class="n">clonotpye</span> <span class="o">=</span> <span class="n">clonotypes</span><span class="p">[</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">clonotype</span><span class="p">]</span>
<span class="n">lineage_tree</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;lineage_tree_files&quot;</span><span class="p">,</span> <span class="n">clonotype</span><span class="o">=</span><span class="n">clonotype</span><span class="p">)</span>

<span class="c1">#an additional label to append to the file names can be optional provided</span>
<span class="n">lineage_tree</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;lineage_tree_files&quot;</span><span class="p">,</span> <span class="n">clonotype</span><span class="o">=</span><span class="n">clonotype</span><span class="p">,</span>   <span class="n">tree_label</span><span class="o">=</span><span class="s2">&quot;best&quot;</span><span class="p">)</span>
</code></pre></div>
<p>The <a href="../../api/lineagetree/">LineageTree</a> class also provides the ability to perform preorder, postorder traversals of the nodes or to iterate through all the nodes or edges in the lineage tree.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span>
<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>
<span class="c1">#preorder traversal</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">lineage_tree</span><span class="o">.</span><span class="n">preorder_traversal</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Node </span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="se">\n</span><span class="s2">BCR Sequence:</span><span class="si">{</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">n</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">Isotype:</span><span class="si">{</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">isotypes</span><span class="p">[</span><span class="n">isotypes</span><span class="p">[</span><span class="n">n</span><span class="p">]]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1">#postoder traversal</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">lineage_tree</span><span class="o">.</span><span class="n">postorder_traversal</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="se">\n</span><span class="s2">BCR Sequence</span><span class="si">{</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">n</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">Isotype</span><span class="si">{</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">isotypes</span><span class="p">[</span><span class="n">isotypes</span><span class="p">[</span><span class="n">n</span><span class="p">]]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1">#iterate over nodes</span>
<span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">lineage_tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="se">\n</span><span class="s2">BCR Sequence</span><span class="si">{</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">sequences</span><span class="p">[</span><span class="n">n</span><span class="p">]</span><span class="si">}</span><span class="se">\n</span><span class="s2">Isotype</span><span class="si">{</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">isotypes</span><span class="p">[</span><span class="n">isotypes</span><span class="p">[</span><span class="n">n</span><span class="p">]]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1">#iterate over edges</span>
<span class="k">for</span> <span class="n">u</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">lineage_tree</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">u</span><span class="si">}</span><span class="s2">-&gt;</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<p>Lastly, you can query the parent, children, leaf status, root status of a specified node:</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span>
<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>

<span class="c1">#preorder traversal</span>
<span class="n">nodes</span> <span class="o">=</span> <span class="n">lineage_tree</span><span class="o">.</span><span class="n">nodes</span><span class="p">()</span>
<span class="n">n</span> <span class="o">=</span> <span class="n">nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="c1">#get parent of node n</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

<span class="c1">#get children of node n</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

<span class="c1">#check if node n is a leaf</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

<span class="c1">#check if node n is the root</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">is_root</span><span class="p">(</span><span class="n">n</span><span class="p">))</span>

<span class="c1">#get the leafset </span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">get_leafs</span><span class="p">())</span>

<span class="c1">#get a dictionary containing the parent of every node</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage_tree</span><span class="o">.</span><span class="n">get_parents</span><span class="p">())</span>
</code></pre></div>
<h3 id="functionality-of-a-lineagetreelist">Functionality of a LineageTreeList<a class="headerlink" href="#functionality-of-a-lineagetreelist" title="Permanent link">&para;</a></h3>
<p>The <a href="../../api/lineagetreelist/">LineageTreeList</a> is an extension of list, which provides additional functionality for organizing a list of <a href="../../api/lineagetree/">LineageTree</a> objects. </p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree</span><span class="p">,</span> <span class="n">LineageTreeList</span>

<span class="n">lt_list</span> <span class="o">=</span> <span class="n">LineageTreeList</span><span class="p">()</span>
<span class="n">lt_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">lineage_tree</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lt_list</span><span class="p">)</span>
</code></pre></div>
<p>The <a href="../../api/lineagetreelist/">LineageTreeList</a> class provides functionality to find the optimal or all optimal <a href="../../api/lineagetree/">LineageTree</a>  in the list or randomly sample one. </p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree_list</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">lineage_tree_list</span><span class="p">))</span>
<span class="c1">#if there are multiple optimal solutions, the first in the list is returned</span>
<span class="n">best_score</span><span class="p">,</span> <span class="n">best_tree</span> <span class="o">=</span> <span class="n">lineage_tree_list</span><span class="o">.</span><span class="n">find_best_tree</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">best_score</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">best_tree</span><span class="p">)</span>

<span class="n">best_score</span><span class="p">,</span> <span class="n">all_best</span>  <span class="o">=</span> <span class="n">lineage_tree_list</span><span class="o">.</span><span class="n">find_all_best_trees</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">best_score</span><span class="p">)</span>

<span class="n">random_score</span><span class="p">,</span> <span class="n">random_tree</span> <span class="o">=</span> <span class="n">lineage_tree_list</span><span class="o">.</span><span class="n">sample_best_tree</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">random_tree</span><span class="p">)</span>
</code></pre></div>
<p>In addition, the class provides a wrapper to the <code>write</code> function in <a href="../../api/lineagetree/">LineageTree</a> to write all the lineage tree files to disk.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree_list</span>


<span class="n">lineage_tree_list</span><span class="o">.</span><span class="n">write_all</span><span class="p">(</span><span class="n">outpath</span><span class="o">=</span><span class="s2">&quot;all_trees&quot;</span><span class="p">)</span>

<span class="c1">#or to utilize isotype labels for the isotype files instead of numerical encoding</span>
<span class="n">isotypes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHM&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span> <span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span><span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span><span class="s1">&#39;IGHE&#39;</span><span class="p">,</span><span class="s1">&#39;IGHA2&#39;</span><span class="p">]</span>

<span class="n">lineage_tree_list</span><span class="o">.</span><span class="n">write_all</span><span class="p">(</span><span class="n">outpath</span><span class="o">=</span><span class="s2">&quot;all_trees&quot;</span><span class="p">,</span> <span class="n">isotype_encoding</span><span class="o">=</span><span class="n">isotypes</span><span class="p">)</span>
</code></pre></div>
<p>Lastly, a CSV file with the objective values of each <a href="../../api/lineagetree/">LineageTree</a> in the list can be written.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">tribal</span> <span class="kn">import</span> <span class="n">lineage_tree_list</span>
<span class="n">lineage_tree_list</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;objectives.csv&quot;</span><span class="p">)</span>
</code></pre></div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.top", "navigation.path", "toc.integrate", "toc.collapsible", "search.suggest", "search.highlight", "content.code.annotation", "content.code.copy", "content.tabs.link"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
    
  </body>
</html>